### PR TITLE
[FLOSS T1] Fix missing array meta-key in the csvstorage plugin

### LINKED ISSUE #3369: [FLOSS H1] CSV Storage: Plugin Should Add Array Metadata
# Problem Description

Currently the [`csvstorage` plugin](https://www.libelektra.org/plugins/csvstorage) does not add the [`array` meta key](https://github.com/ElektraInitiative/libelektra/blob/master/doc/tutorials/arrays.md) to array parents. As a consequence of this behavior other plugins, such as [Directory Value](https://www.libelektra.org/plugins/directoryvalue), [will not work as expected](https://github.com/ElektraInitiative/libelektra/pull/3357#discussion_r386127533). 

# Steps to Reproduce the Problem

```sh
kdb mount config.csv user:/tests/csvstorage csvstorage
printf 'one,two,three\nfour,five,six\n' > `kdb file user:/tests/csvstorage`
kdb ls user:/tests/csvstorage
#> user/tests/csvstorage/#0
#> user/tests/csvstorage/#0/#0
#> user/tests/csvstorage/#0/#1
#> user/tests/csvstorage/#0/#2
#> user/tests/csvstorage/#1
#> user/tests/csvstorage/#1/#0
#> user/tests/csvstorage/#1/#1
#> user/tests/csvstorage/#1/#2

# The plugin should add the array meta key to array parents, but does not.
kdb meta-get user:/tests/csvstorage/#0 array
# RET: 2
```

# Expected Result

The command `meta-get` should print the value `#2` (the [index of the last array element](https://github.com/ElektraInitiative/libelektra/blob/master/doc/decisions/array.md#decision)) and finish successfully (return code `0`).

# Actual Result

The command `meta-get` prints nothing and returns the unsuccessfully (return code `2`).

### PR DESCRIPTION
**Pull request by muskater**
## Basics
This fixes #3369 . We have added the missing `array` meta-key type to parent key. We also added test cases as markdown shell recorder and as a ctest.

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

## FLOSS
- We found the bug from @muskater 's H2 issues and thought  that it would be good fit.
- The tests run on Jenkins as well as the Mac OS X / FreeBSD Cirrus CI
- All 3 tried to reproduce the issue manually and then we pooled our findings together and wrote the test case in a shared manner. Then we reviewed the source code together via Screen Sharing and agreed upon a solution together. The act of writing the 2 lines was more of a symbolic act, after we determined the solution together.

### CODE REVIEW

File: tests/shell/shell_recorder/replay_tests/csv_metakey.epf [hunk 1]
```diff
@@ -0,0 +1,30 @@
+CMD: kdb mount config.csv user:/tests/csvstorage csvstorage
```
Comment (markus2330) [inline tests/shell/shell_recorder/replay_tests/csv_metakey.epf:1]: Thank you so much! Did you use on purpose the non-markdown variant of the shell recorder? See https://github.com/ElektraInitiative/libelektra/blob/master/tests/shell/shell_recorder/tutorial_wrapper/README.md for markdown variant.

Comment (lawli3t) [reply tests/shell/shell_recorder/replay_tests/csv_metakey.epf:1]: We were not aware of the markdown variant

**Comment by muskater** [comment]
jenkins build libelektra please

**Comment by 4ydan** [comment]
```
kdb mount config.csv user:/tests/csvstorage csvstorage


ERROR - RET:

Return value “7” does not match “0”


kdb set user:/tests/csvstorage/test test

printf 'one,two,three\nfour,five,six\n' > `kdb file user:/tests/csvstorage`

kdb ls user:/tests/csvstorage


ERROR - RET:

Return value “5” does not match “0”


kdb meta-get user:/tests/csvstorage/#0 array


ERROR - RET:

Return value “5” does not match “0”


kdb rm -r user:/tests/csvstorage


ERROR - RET:

Return value “5” does not match “0”


kdb umount user:/tests/csvstorage


ERROR - RET:

Return value “1” does not match “0”


=======================================

Replay test failed, protocols differ
```

We cant explain to ourselves why our new testcase fails with RET 7 "Unknown errors, wrong exceptions thrown." 
Following the docker tutorial to run all tests, the tests pass on our machines.

File: tests/shell/shell_recorder/replay_tests/csv_metakey.epf (hunk 1 above)
Comment (muskater) [reply tests/shell/shell_recorder/replay_tests/csv_metakey.epf]: We have now converted the test cases into the Markdown variant

**Comment by markus2330** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
> It seems like the build server does not have the csv plugin while running our custom shellrecorded script.

Jenkins fails because of a spelling mistake. Where do you see the failure because of the missing plugin?

**Review by JakobWonisch — APPROVED** [review]
[FLOSS T1] Review

It seems you have very thoroughly worked on this issue. The problem is fixed with the lines you added and test cases automatically checking whether that is the case are also in place.

All in all, this looks good to me!

**Pull request merged**